src/core/expression/predicate.visitor.js

```diff
@@ -40,22 +40,19 @@ export class PredicateVisitor extends Visitor {
 		const assert = this.assertFactory(name);
 		const map = new Set();
 
-		let rCastAs;
+		const rt = getType(isArray(r) ? r[0] : r);
+		const parse = parseFactory(rt);
+
 		if (isArray(r)) {
 			if (r.length) {
-				rCastAs = castAsFactory(r[0]);
 				r.forEach(x => map.add('' + x));
 			} else {
-				rCastAs = identity;
+				parse = identity;
 			}
-		} else {
-			rCastAs = castAsFactory(r);
 		}
 
-		const equals = assert.equals;
-		const isNull = assert.isNull;
-		const lessThan = assert.lessThan;
-		const lessThanOrEquals = (x, y) => equals(x, y) || lessThan(x, y);
+		const { equals, isNull, lessThan } = assert;
+		const lessThanOrEquals = (x, y) => equals(parse(x), parse(y)) || lessThan(x, y);
```

Comment:
why in one case we use parse and in another not?

Reply:
Indeed, it slightly weird step which obviously was bound with parsing of `r` ) Fixed in next commit. Thanks